File: tests/src/org.apache.hadoop.thirdparty/hadoop-shaded-protobuf_3_7/1.1.1/src/test/java/org_apache_hadoop_thirdparty/hadoop_shaded_protobuf_3_7/ExtensionRegistryLiteTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_apache_hadoop_thirdparty.hadoop_shaded_protobuf_3_7;

import static org.assertj.core.api.Assertions.assertThat;

import org.apache.hadoop.thirdparty.protobuf.Descriptors;
import org.apache.hadoop.thirdparty.protobuf.Extension;
import org.apache.hadoop.thirdparty.protobuf.ExtensionRegistry;
import org.apache.hadoop.thirdparty.protobuf.ExtensionRegistryLite;
import org.apache.hadoop.thirdparty.protobuf.MessageLite;
import org.apache.hadoop.thirdparty.protobuf.WireFormat;
import org.junit.jupiter.api.Test;

public class ExtensionRegistryLiteTest {
    @Test
    void addDispatchesExtensionLiteToFullRegistryImplementation() {
        ExtensionRegistryLite registry = ExtensionRegistryLite.newInstance();
        RecordingProto1Extension extension = new RecordingProto1Extension();

        registry.add(extension);

        assertThat(registry).isInstanceOf(ExtensionRegistry.class);
        assertThat(extension.extensionTypeRequests()).isPositive();
    }

    private static final class RecordingProto1Extension extends Extension<MessageLite, Integer> {
        private int extensionTypeRequests;

        @Override
        public Descriptors.FieldDescriptor getDescriptor() {
            return null;
        }

        @Override
        protected ExtensionType getExtensionType() {
            extensionTypeRequests++;
            return ExtensionType.PROTO1;
        }

        @Override
        protected Object fromReflectionType(Object value) {
            return value;
        }

        @Override
        protected Object singularFromReflectionType(Object value) {
            return value;
        }

        @Override
        protected Object toReflectionType(Object value) {
            return value;
        }

        @Override
        protected Object singularToReflectionType(Object value) {
            return value;
        }

        @Override
        public int getNumber() {
            return 321;
        }

        @Override
        public WireFormat.FieldType getLiteType() {
            return WireFormat.FieldType.INT32;
        }

        @Override
        public boolean isRepeated() {
            return false;
        }

        @Override
        public Integer getDefaultValue() {
            return 0;
        }

        @Override
        public MessageLite getMessageDefaultInstance() {
            return null;
        }

        int extensionTypeRequests() {
            return extensionTypeRequests;
        }
    }
}

File: tests/src/org.apache.pekko/pekko-protobuf-v3_2.13/1.2.0/src/test/scala/org_apache_pekko/pekko_protobuf_v3_2_13/ExtensionRegistryLiteTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package org_apache_pekko.pekko_protobuf_v3_2_13

import org.apache.pekko.protobufv3.internal.Descriptors
import org.apache.pekko.protobufv3.internal.Extension
import org.apache.pekko.protobufv3.internal.ExtensionLite
import org.apache.pekko.protobufv3.internal.ExtensionRegistry
import org.apache.pekko.protobufv3.internal.ExtensionRegistryLite
import org.apache.pekko.protobufv3.internal.Message
import org.apache.pekko.protobufv3.internal.MessageLite
import org.apache.pekko.protobufv3.internal.WireFormat
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ExtensionRegistryLiteTest {
  @Test
  def addExtensionLiteDispatchesFullExtensionThroughRegistryReflection(): Unit = {
    val registry: ExtensionRegistryLite = ExtensionRegistry.newInstance()
    val extension: Proto1Extension = new Proto1Extension
    val liteExtension: ExtensionLite[_ <: MessageLite, String] = extension

    registry.add(liteExtension)

    assertThat(registry).isInstanceOf(classOf[ExtensionRegistry])
    assertThat(extension.extensionTypeLookupCount).isPositive
  }

  private final class Proto1Extension extends Extension[Message, String] {
    private var extensionTypeLookups: Int = 0

    def extensionTypeLookupCount: Int = extensionTypeLookups

    override def getNumber: Int = 12345

    override def getLiteType: WireFormat.FieldType = WireFormat.FieldType.STRING

    override def isRepeated: Boolean = false

    override def getDefaultValue: String = "default"

    override def getMessageDefaultInstance: Message = null

    override def getDescriptor: Descriptors.FieldDescriptor =
      throw new AssertionError("PROTO1 extensions do not require descriptors when registered")

    override protected def getExtensionType: Extension.ExtensionType = {
      extensionTypeLookups += 1
      Extension.ExtensionType.PROTO1
    }

    override protected def fromReflectionType(value: AnyRef): AnyRef = value

    override protected def singularFromReflectionType(value: AnyRef): AnyRef = value

    override protected def toReflectionType(value: AnyRef): AnyRef = value

    override protected def singularToReflectionType(value: AnyRef): AnyRef = value
  }
}
